### DISCUSSION TITLE
[Contract Verification] Cannot link proxy contract with its implementation on explorer

### DISCUSSION THREAD

**Discussion by smusialpc** [zkSync-Community-Hub/zksync-developers#76] [Contract Verification] [answered]
### Environment

Testnet

### Issue Type

- [ ] Plugin
- [X] Block Explorer

### Contract Address

0x0c3437C2F8be296927b2A152C42541d6414C8Ba9

### Compiler Type

Multi-part contract

### zkSolc Version

v1.3.13

### Solc Version

0.8.18

### Contract Name

TestZkSync

### Contract Code

pragma solidity ^0.8.18;

import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";

contract TestZkSync is Initializable {
    string public name;
    uint256 public age;

    /// @custom:oz-upgrades-unsafe-allow constructor
    constructor() {
        _disableInitializers();
    }

    function initialize(string memory name_) public initializer {
        name = name_;
    }

    function setName(string memory newName) external {
        name = newName;
    }

    function setAge(uint256 newAge) external {
        age = newAge;
    }

    function getBlockNumber() external view returns (uint256) {
        return block.number;
    }

    function getBlockTimestamp() external view returns (uint256) {
        return block.timestamp;
    }

    uint256[48] private __gap;
}

### Constructor Arguments

0x

### Hardhat Verify Plugin Version

0.2.0

### Additional Details

I am not able to link proxy contracts with their implementation on zkSync explorer.
Etherscan has option 'Is this a proxy?' in Contract -> Code tab which allows using implementation ABI on proxy contract. I don't see similar feature in zkSync (at least on testnet). Without such link I can't call transactions using explorer even if the contracts are verified.

I deployed contracts (transparent proxy + implementation) and verified them according to hardhat-zksync-upgradable and hardhat-zksync-verify docs. Now I have both contracts verified, but still can't call implementation ABI from proxy.

Proxy: 0x0c3437C2F8be296927b2A152C42541d6414C8Ba9
Impl: 0xF5C36dd93ee1034756fA140BE746c6C5B7BC367a

**Comment by uF4No** [comment, marked as answer]
Thanks for the feedback. As you said, currently the explorer does not link the proxy and implementation contracts, so you cannot read/write directly from the explorer and pointing to the active implementation contract. That said, this was already in our pipeline so it should be added to the explorer in the upcoming weeks.

**Reply by githubdoramon** [reply]
Little bird told me this is coming quite soon

**Reply by bxpana** [reply]
There's now a banner in proxy contracts on the contract verification on the block explorer with a link to the implementation contract 
![image](https://github.com/zkSync-Community-Hub/zksync-developers/assets/42230936/5f9b27e3-546f-4278-ab18-5b31c21e6917)

**Comment by xiaoch05** [comment]
> Thanks for the feedback. As you said, currently the explorer does not link the proxy and implementation contracts, so you cannot read/write directly from the explorer and pointing to the active implementation contract. That said, this was already in our pipeline so it should be added to the explorer in the upcoming weeks.

Looking forward to the progress of this feature.

**Comment by guarinojoe** [comment]
To link proxy contracts with their implementations on zkSync explorer, you may need to manually specify the implementation address in your proxy contract's constructor or a dedicated function. This linking process ensures that the proxy contract delegates calls to the correct implementation. Additionally, zkSync may not have a built-in feature like Etherscan's 'Is this a proxy?' option, so manual linking is often required. Ensure that your proxy contract is correctly configured to delegate calls to the implementation contract.

**Comment by beckyhaynes** [comment]
It seems that you are encountering an issue with linking proxy contracts with their implementation on zkSync Explorer. Unlike Etherscan, zkSync Explorer currently does not have a feature to indicate whether a contract is a proxy or not, which makes it challenging to use the implementation ABI on proxy contracts.

To address this issue, you may consider the following steps:

Request Feature: Reach out to zkSync Explorer's development team and request the implementation of a feature similar to Etherscan's "Is this a proxy?" option. This would enable users to link proxy contracts with their implementations more easily.

Alternative Solutions: In the absence of direct support on zkSync Explorer, you can consider alternative solutions such as maintaining documentation or metadata indicating the relationship between proxy contracts and their implementations. This can help developers understand the structure of your contracts and how to interact with them.

Community Discussion: Engage with the zkSync community, forums, or social media channels to discuss this issue and explore if other users have encountered similar challenges or have devised workarounds.

Manual Verification: Until a dedicated feature is implemented, you may need to manually manage the connection between proxy contracts and their implementations, ensuring that users are aware of how to interact with them properly.

By taking these steps and actively engaging with the zkSync community and development team, you can work towards finding a solution or improving the usability of proxy contracts on zkSync Explorer.

**Comment by hanjv** [comment]
To link proxy contracts with their implementation on zkSync explorer, ensure that both the proxy contract and its implementation contract are verified. Once verified, check if the explorer supports displaying proxy contracts and their linked implementations. If not, consider reaching out to the zkSync team for assistance or clarification on this matter.

**Comment by jwitte3** [comment]
To link proxy contracts with their implementation on zkSync explorer:
   - Ensure both the proxy and implementation contracts are verified.
   - Look for an option similar to 'Is this a proxy?' in zkSync explorer to associate the proxy with its implementation ABI.
   - Without this link, calling transactions via the explorer may not be possible even if contracts are verified.
